// gatsby/dist/schema/type-conflict-reporter.js
"use strict";

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _ = require(`lodash`);
var report = require(`gatsby-cli/lib/reporter`);
var typeOf = require(`type-of`);
var util = require(`util`);

var _require = require(`./node-tracking`),
    findRootNodeAncestor = _require.findRootNodeAncestor;

var isNodeWithDescription = function isNodeWithDescription(node) {
  return node && node.internal && node.internal.description;
};

var findNodeDescription = function findNodeDescription(obj) {
  if (obj) {
    var node = findRootNodeAncestor(obj, isNodeWithDescription);
    if (isNodeWithDescription(node)) {
      return node.internal.description;
    }
  }
  return ``;
};

var formatValue = function formatValue(value) {
  if (!_.isArray(value)) {
    return util.inspect(value, {
      colors: true,
      depth: 0,
      breakLength: Infinity
    });
  }

  var wasElipsisLast = false;
  var usedTypes = [];
  var output = [];

  value.forEach(function (item) {
    var type = typeOf(item);
    if (usedTypes.indexOf(type) !== -1) {
      if (!wasElipsisLast) {
        output.push(`...`);
        wasElipsisLast = true;
      }
    } else {
      output.push(formatValue(item));
      wasElipsisLast = false;
      usedTypes.push(type);
    }
  });

  return `[ ${output.join(`, `)} ]`;
};

var TypeConflictEntry = function () {
  function TypeConflictEntry(selector) {
    (0, _classCallCheck3.default)(this, TypeConflictEntry);

    this.selector = selector;
    this.types = {};
  }

  TypeConflictEntry.prototype.addExample = function addExample(_ref) {
    var value = _ref.value,
        type = _ref.type,
        parent = _ref.parent;

    this.types[type] = {
      value,
      description: findNodeDescription(parent)
    };
  };

  TypeConflictEntry.prototype.printEntry = function printEntry() {
    var sortedByTypeName = _.sortBy(_.entries(this.types), function (_ref2) {
      var typeName = _ref2[0],
          value = _ref2[1];
      return typeName;
    });

    report.log(`${this.selector}:${sortedByTypeName.map(function (_ref3) {
      var typeName = _ref3[0],
          _ref3$ = _ref3[1],
          value = _ref3$.value,
          description = _ref3$.description;
      return `\n - type: ${typeName}\n   value: ${formatValue(value)}${description && `\n   source: ${description}`}`;
    }).join(``)}`);
  };

  return TypeConflictEntry;
}();

var TypeConflictReporter = function () {
  function TypeConflictReporter() {
    (0, _classCallCheck3.default)(this, TypeConflictReporter);

    this.clearConflicts();
  }

  TypeConflictReporter.prototype.clearConflicts = function clearConflicts() {
    this.entries = {};
  };

  TypeConflictReporter.prototype.getFromSelector = function getFromSelector(selector) {
    if (this.entries[selector]) {
      return this.entries[selector];
    }

    var dataEntry = new TypeConflictEntry(selector);
    this.entries[selector] = dataEntry;
    return dataEntry;
  };

  TypeConflictReporter.prototype.addConflict = function addConflict(selector, examples) {
    if (selector.substring(0, 11) === `SitePlugin.`) {
      // Don't store and print out type conflicts in plugins.
      // This is out of user control so he can't do anything
      // to hide those.
      return;
    }

    var entry = this.getFromSelector(selector);
    examples.filter(function (example) {
      return example.value != null;
    }).forEach(function (example) {
      return entry.addExample(example);
    });
  };

  TypeConflictReporter.prototype.printConflicts = function printConflicts() {
    var entries = _.values(this.entries);
    if (entries.length > 0) {
      report.warn(`There are conflicting field types in your data. GraphQL schema will omit those fields.`);
      entries.forEach(function (entry) {
        return entry.printEntry();
      });
    }
  };

  return TypeConflictReporter;
}();

var typeConflictReporter = new TypeConflictReporter();

var printConflicts = function printConflicts() {
  typeConflictReporter.printConflicts();
};

module.exports = { typeConflictReporter, printConflicts, TypeConflictEntry };
